tensorflow/c/experimental/filesystem/plugins/hadoop/hadoop_filesystem.cc

```diff
@@ -200,11 +201,17 @@ hdfsFS Connect(LibHDFS* libhdfs, const std::string& path, TF_Status* status) {
     SplitArchiveNameAndPath(&path_har, &namenode, status);
     if (TF_GetCode(status) != TF_OK) return nullptr;
     libhdfs->hdfsBuilderSetNameNode(builder, namenode.c_str());
+    cacheKey += namenode;
   } else {
     libhdfs->hdfsBuilderSetNameNode(
         builder, namenode.empty() ? "default" : namenode.c_str());
+    cacheKey += namenode;
   }
-  auto fs = libhdfs->hdfsBuilderConnect(builder);
+  auto connection_cache = hadoop_file->connection_cache;
+  if (connection_cache.find(cacheKey) == connection_cache.end()) {
+    connection_cache[cacheKey] = libhdfs->hdfsBuilderConnect(builder);
+  }
```

Comment:
It is a little bit tricky here. Since `hadoop_file->connection_cache` isn't a pointer, `auto connection_cache = hadoop_file->connection_cache;` is a copy-by-value instead of taking address. We want either `hadoop_file->connection_cache.find()` or `auto connection_cache = &hadoop_file->connection_cache;`

And you forgot the `Cleanup` function in the filesystem.

Reply:
@vnvo2409
Can you please review it again?

Reply:
Look good to me. But I don't have the write access so we will have to wait for @mihaimaruseac